command line SF_OP improvements

HeadLessHUN opened the pull request:
1. -SF_OP command now supports MIN/DISP_MIN/SAT_MIN/N_SIGMA_MIN/MAX/DISP_MAX/SAT_MAX/N_SIGMA_MAX as value
 2. add new option to SF_OP {sf} {operation} {value} -NOT_IN_PLACE, to create new scalar field during the operation

reference:
https://www.cloudcompare.org/forum/viewtopic.php?f=9&t=6784

dgirardeau commented:
I'm not sure to understand why you  need to add the sub-option -NOT_IN_PLACE at the end? 🤔

dgirardeau commented on qCC/ccCommandLineCommands.cpp:
```diff
@@ -2743,6 +2744,29 @@ static std::pair<ScalarType, ScalarType> GetSFRange(const CCCoreLib::ScalarField
+static ScalarType GetSFValue(ccPointCloud* pc, int sfIndex, ScalarType value, USE_SPECIAL_SF_VALUE useVal)
+{
+	//should be handled way before this point this is just safety
+	if (pc)
```
another option is to use a reference to a cloud instead of a pointer

HeadLessHUN replied:
so if i define like the folowwing the pc check is not necessary?

static ScalarType GetSFValue(ccPointCloud pc, int sfIndex, ScalarType value, USE_SPECIAL_SF_VALUE useVal)

and i call it with dereferenced value?

&desc.pc

dgirardeau replied:
Yes, but it should be static ScalarType GetSFValue(const ccPointCloud& pc, int sfIndex, ScalarType value, USE_SPECIAL_SF_VALUE useVal)

(to avoid a copy of the cloud)

dgirardeau commented on qCC/ccCommandLineCommands.cpp:
```diff
@@ -5860,13 +5884,19 @@ bool CommandSFOperation::process(ccCommandLineInterface& cmd)
 	
 	//read scalar value
-	double value = 1.0;
+	ScalarType value = 1.0;
```
```suggestion
	ScalarType value = static_cast<ScalarType>(1.0);
```

dgirardeau commented on qCC/ccCommandLineCommands.cpp:
```diff
@@ -5875,7 +5905,19 @@ bool CommandSFOperation::process(ccCommandLineInterface& cmd)
+	bool inPlace = true;
+	if (!cmd.arguments().empty())
+	{
+		if (cmd.IsCommand(cmd.arguments().front(),COMMAND_SF_OP_NOT_IN_PLACE))
```
I would have placed it first (even if optional)

HeadLessHUN replied:
i'll do a bigger change and replace the parameter loading to a while loop similar to other commands. So it can be placed anywhere.

HeadLessHUN replied:
On second thought it is not gonna work in this case. I'll put this option to the front as optional.

dgirardeau commented on qCC/ccCommandLineCommands.cpp:
```diff
@@ -5875,7 +5905,19 @@ bool CommandSFOperation::process(ccCommandLineInterface& cmd)
 	}
-	
+
+	//in place modifier, to keep old commands intact we should keep it in place by default. However it makes the command line inconsistence, because the SF_ARITHMETIC works the other way.
```
```suggestion
	//in place modifier, to keep old commands intact we should keep it in place by default. However it makes the command line inconsistent, because the SF_ARITHMETIC works the other way.
```

dgirardeau reviewed (dismissed):
Thanks for the update

dgirardeau commented:
Can you wait before merging that? There's another PR that I'd like to merge first (the next 2.13.1 version)

dgirardeau reviewed (commented):
Let's wait

HeadLessHUN commented:
> Can you wait before merging that? There's another PR that I'd like to merge first (the next 2.13.1 version)

It is not important for me take your time :) i'm more interested in the color gaussian filtering, but i'm patient :)

dgirardeau commented:
Can you rebase this PR on master?

HeadLessHUN commented:
tomorrow :)

HeadLessHUN commented:
i think i've rebased it correctly everything runs fine, i also merged in a little change for CSF plugin CLI. Not tested that part but i don't see why it won't work.

fixed changelog :)

dgirardeau commented on CHANGELOG.md:
```diff
@@ -34,8 +33,19 @@ New Feature:
 				- only used when the filter is applied to RGB colors
+			- -SIGMA {sigma}
+					- optional
+					- nearest neighbours calculated with a radius of 3*sigma, if not set cloud compare will calculate a default value.
```
```suggestion
					- nearest neighbours extracted with a radius of 3*sigma. If not set, CloudCompare will calculate a default value.
```

dgirardeau commented on CHANGELOG.md:
```diff
@@ -34,8 +33,19 @@ New Feature:
+					- nearest neighbours calculated with a radius of 3*sigma, if not set cloud compare will calculate a default value.
+			- -SIGMA_SF {sigma_sf}
+					- optional, only used when bilateral filter applied
+		- New option to SF_OP -NOT_IN_PLACE {sf} {operation} {value}, to create new scalar field during the operation.
```
```suggestion
		- New SF_OP suboption: -NOT_IN_PLACE
			- to create a new scalar field during the operation
```

dgirardeau commented on CHANGELOG.md:
```diff
@@ -34,8 +33,19 @@ New Feature:
+
+Improvements:
+	- Command line:
+		- -SF_OP command now supports MIN/DISP_MIN/SAT_MIN/N_SIGMA_MIN/MAX/DISP_MAX/SAT_MAX/N_SIGMA_MAX as value
```
```suggestion
		- the -SF_OP command now supports MIN/DISP_MIN/SAT_MIN/N_SIGMA_MIN/MAX/DISP_MAX/SAT_MAX/N_SIGMA_MAX as input values
```

dgirardeau commented on CHANGELOG.md:
```diff
@@ -34,8 +33,19 @@ New Feature:
+Improvements:
+	- Command line:
+		- -SF_OP command now supports MIN/DISP_MIN/SAT_MIN/N_SIGMA_MIN/MAX/DISP_MAX/SAT_MAX/N_SIGMA_MAX as value
+		- Rename -CSF command's result clouds to be able to select them later.
```
```suggestion
		- Rename -CSF command's resulting clouds to be able to select them later:
			- {original cloud name} + '_ground_points'
			- {original cloud name} + '_offground_points'
```

dgirardeau commented on CHANGELOG.md:
```diff
@@ -34,8 +33,19 @@ New Feature:
+		- Rename -CSF command's result clouds to be able to select them later.
 
-v2.13.1 (Kharkiv) - (03/20/2024)
+		v2.13.1 (Kharkiv) - (03/20/2024)
```
restore the original format

dgirardeau commented on qCC/ccCommandLineCommands.cpp:
```diff
@@ -5865,11 +5885,23 @@ CommandSFOperation::CommandSFOperation()
 
 bool CommandSFOperation::process(ccCommandLineInterface& cmd)
 {
+
```
remove blank line
